toucan_connectors/redshift/redshift_database_connector.py

```diff
@@ -0,0 +1,352 @@
+
+import pandas as pd
+import redshift_connector
+from pydantic import Field, SecretStr, create_model, root_validator
+from pydantic.types import constr
+
+from toucan_connectors.common import ConnectorStatus
+from toucan_connectors.connection_manager import ConnectionManager
+from toucan_connectors.sql_query_helper import SqlQueryHelper
+from toucan_connectors.toucan_connector import (
+    DataSlice,
+    DataStats,
+    ToucanConnector,
+    ToucanDataSource,
+    strlist_to_enum,
+)
+
+TABLE_QUERY = """SELECT DISTINCT tablename FROM pg_table_def WHERE schemaname = 'public';"""
+
+ORDERED_KEYS = [
+    'type',
+    'name',
+    'host',
+    'port',
+    'cluster_identifier',
+    'db_user',
+    'connect_timeout',
+    'authentication_method',
+    'user',
+    'password',
+    'access_key_id',
+    'secret_access_key',
+    'session_token',
+    'profile',
+    'region',
+]
+
+logger = logging.getLogger(__name__)
+
+redshift_connection_manager = None
+if not redshift_connection_manager:
+    redshift_connection_manager = ConnectionManager(
+        name='redshift', timeout=25, wait=0.2, time_between_clean=10, time_keep_alive=600
+    )
+
+
+class AuthenticationMethod(str, Enum):
+    DB_CREDENTIALS: str = 'db_credentials'
+    AWS_CREDENTIALS: str = 'aws_credentials'
+    AWS_PROFILE: str = 'aws_profile'
+
+
+class AuthenticationMethodError(str, Enum):
+    DB_CREDENTIALS: str = f'User & Password are required for {AuthenticationMethod.DB_CREDENTIALS}'
+    AWS_CREDENTIALS: str = f'AccessKeyId, SecretAccessKey & db_user are required for {AuthenticationMethod.AWS_CREDENTIALS}'
+    AWS_PROFILE: str = f'Profile & db_user are required for {AuthenticationMethod.AWS_PROFILE}'
+    UNKNOWN: str = 'Unknown AuthenticationMethod'
+
+
+class RedshiftDataSource(ToucanDataSource):
+    database: str = Field(..., description='The name of the database you want to query')
+    query: constr(min_length=1) = Field(
+        None,
+        description='You can write a custom query against your '
+        'database here. It will take precedence over '
+        'the table parameter',
+        widget='sql',
+    )
+    table: constr(min_length=1) = Field(
+        None,
+        description='The name of the data table that you want to '
+        'get (equivalent to "SELECT * FROM '
+        'your_table")',
+    )
+
+    def __init__(self, **data):
+        super().__init__(**data)
+        query = data.get('query')
+        table = data.get('table')
+        if query is None and table is None:
+            raise ValueError("'query' or 'table' must be set")
+        elif query is None and table is not None:
+            self.query = f'select * from {table};'
+
+    @classmethod
+    def get_form(cls, connector: 'RedshiftConnector', current_config):
+        constraints = {}
+        with suppress(Exception):
+            if 'database' in current_config:
+                available_tables = connector._retrieve_tables(current_config['database'])
```

Comment:
Pourquoi ici on passe la str `database` alors que la signature de la fonction attend une datasource ? 
`def _retrieve_tables(self, datasource) -> List[str]:`
Le pb sur la récupération des tables dans le formulaire data source vient de là.

Reply:
Yes effectivement, c'est surement un relicat de nos différentes itérations

On fix ça avec le comment précédente